### PR TITLE
Hotfix ETP-3395: Merge develop into main

### PR DESCRIPTION
**Pull request by Luciano-Palacio**

### CODE REVIEW

File: src-test/src/com/etendoerp/metadata/auth/AuthDataTest.java [hunk 1]
```diff
@@ -27,7 +27,8 @@
  * warehouse, and client information.
  */
 @RunWith(MockitoJUnitRunner.class)
-public class AuthDataTest {
+public class
+AuthDataTest {
```
Comment (etendobot) [inline src-test/src/com/etendoerp/metadata/auth/AuthDataTest.java:31]: **Suggestion**

Revert the class declaration to a single line.

```java
public class AuthDataTest {
```

**Rationale**: Standard Java and Etendo coding conventions specify that class declarations should be on a single line unless they are exceptionally long. Splitting the declaration here reduces readability and appears to be an accidental formatting change.

File: src-test/src/com/etendoerp/metadata/http/AttachmentsServletCoverageTest.java [hunk 2]
```diff
@@ -0,0 +1,576 @@
+package com.etendoerp.metadata.http;
+
```
Comment (etendobot) [inline src-test/src/com/etendoerp/metadata/http/AttachmentsServletCoverageTest.java:2]: > [!WARNING]
**⚠️ Blocking Issue**

Add the mandatory Etendo license header at the beginning of the file.

```java
/*\n *************************************************************************\n * The contents of this file are subject to the Etendo License\n * (the \"License\"), you may not use this file except in compliance\n * with the License.\n * You may obtain a copy of the License at\n * https://github.com/etendosoftware/etendo_core/blob/main/legal/Etendo_license.txt\n * Software distributed under the License is distributed on an\n * \"AS IS\" basis, WITHOUT WARRANTY OF ANY KIND, either express or\n * implied. See the License for the specific language governing rights\n * and limitations under the License.\n * All portions are Copyright (C) 2021-2024 FUTIT SERVICES, S.L\n * All Rights Reserved.\n * Contributor(s): Futit Services S.L.\n *************************************************************************\n */\npackage com.etendoerp.metadata.http;
```

**Rationale**: Mandatory for compliance with Etendo development standards. New files in the com.etendoerp.* namespace must include the license header at the top of the file.

File: src-test/src/com/etendoerp/metadata/builders/FieldBuilderAdditionalTest.java [hunk 3]
```diff
@@ -0,0 +1,858 @@
+package com.etendoerp.metadata.builders;
+
```
Comment (etendobot) [inline src-test/src/com/etendoerp/metadata/builders/FieldBuilderAdditionalTest.java:2]: > [!WARNING]
**⚠️ Blocking Issue**

Add the mandatory Etendo license header at the top of the new file.

```java
/*
 *************************************************************************
 * The contents of this file are subject to the Etendo License
 * (the "License"), you may not use this file except in compliance
 * with the License.
 * You may obtain a copy of the License at
 * https://github.com/etendosoftware/etendo_core/blob/main/legal/Etendo_license.txt
 * Software distributed under the License is distributed on an
 * "AS IS" basis, WITHOUT WARRANTY OF ANY KIND, either express or
 * implied. See the License for the specific language governing rights
 * and limitations under the License.
 * All portions are Copyright (C) 2021-2026 FUTIT SERVICES, S.L
 * All Rights Reserved.
 * Contributor(s): Futit Services S.L.
 *************************************************************************
 */
package com.etendoerp.metadata.builders;
```

**Rationale**: Every new Java class in a com.etendoerp.* module must include the standard Etendo license header to comply with legal and project standards.

File: src-test/src/com/etendoerp/metadata/builders/FieldBuilderAdditionalTest.java [hunk 4]
```diff
@@ -0,0 +1,858 @@
...
+import static org.junit.jupiter.api.Assertions.assertNull;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.mockito.Answers.CALLS_REAL_METHODS;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.mockConstruction;
+import static org.mockito.Mockito.mockStatic;
```
Comment (etendobot) [inline src-test/src/com/etendoerp/metadata/builders/FieldBuilderAdditionalTest.java:14]: **Suggestion**

Remove unused static import for 'mockConstruction'.

```java
// Remove unused import
```

**Rationale**: Unused imports clutter the code and should be removed to maintain cleanliness.

File: src-test/src/com/etendoerp/metadata/builders/FieldBuilderAdditionalTest.java [hunk 5]
```diff
@@ -0,0 +1,858 @@
...
+
+import org.codehaus.jettison.json.JSONArray;
+import org.codehaus.jettison.json.JSONException;
+import org.codehaus.jettison.json.JSONObject;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.Mock;
+import org.mockito.MockedConstruction;
+import org.mockito.MockedStatic;
```
Comment (etendobot) [inline src-test/src/com/etendoerp/metadata/builders/FieldBuilderAdditionalTest.java:29]: **Suggestion**

Remove unused import 'MockedConstruction'.

```java
// Remove unused import
```

**Rationale**: Unused imports should be removed to keep the source file clean.

File: src-test/src/com/etendoerp/metadata/builders/FieldBuilderAdditionalTest.java [hunk 6]
```diff
@@ -0,0 +1,858 @@
...
+import org.mockito.quality.Strictness;
+import org.openbravo.base.model.Entity;
+import org.openbravo.base.model.ModelProvider;
+import org.openbravo.base.model.Property;
+import org.openbravo.base.model.domaintype.DomainType;
+import org.openbravo.base.model.domaintype.ForeignKeyDomainType;
+import org.openbravo.base.model.domaintype.PrimitiveDomainType;
+import org.openbravo.dal.core.DalUtil;
+import org.openbravo.dal.core.OBContext;
+import org.openbravo.model.ad.datamodel.Column;
```
Comment (etendobot) [inline src-test/src/com/etendoerp/metadata/builders/FieldBuilderAdditionalTest.java:41]: **Suggestion**

Remove unused import 'OBContext'.

```java
// Remove unused import
```

**Rationale**: The OBContext class is imported but not used in the test class.

File: src-test/src/com/etendoerp/metadata/builders/FieldBuilderAdditionalTest.java [hunk 7]
```diff
@@ -0,0 +1,858 @@
...
+import org.openbravo.model.ad.datamodel.Table;
+import org.openbravo.model.ad.domain.Reference;
+import org.openbravo.model.ad.system.Language;
+import org.openbravo.model.ad.ui.Field;
+import org.openbravo.model.ad.ui.Tab;
+import org.openbravo.service.datasource.DataSource;
+import org.openbravo.service.datasource.DatasourceField;
+import org.openbravo.service.json.DataToJsonConverter;
+import org.openbravo.service.json.JsonConstants;
```
Comment (etendobot) [inline src-test/src/com/etendoerp/metadata/builders/FieldBuilderAdditionalTest.java:50]: **Suggestion**

Remove unused import 'DataToJsonConverter'.

```java
// Remove unused import
```

**Rationale**: The DataToJsonConverter class is not used in the current test suite.

File: src-test/src/com/etendoerp/metadata/builders/FieldBuilderAdditionalTest.java [hunk 8]
```diff
@@ -0,0 +1,858 @@
...
+        method.setAccessible(true);
+        return method;
+    }
+
+    private void assertNullSelectorFieldDomainType(SelectorField selectorField) {
+        try {
+            Method method = getSelectorFieldDomainTypeMethod();
+            assertNull(invokeSelectorFieldDomainType(method, selectorField));
+        } catch (NoSuchMethodException e) {
+            assertNull(null);
```
Comment (etendobot) [inline src-test/src/com/etendoerp/metadata/builders/FieldBuilderAdditionalTest.java:192]: **Suggestion**

Consider failing the test explicitly if the reflected method is not found, rather than using a redundant assertNull(null).

```java
        } catch (NoSuchMethodException e) {
            org.junit.jupiter.api.Assertions.fail("Method getDomainType not found", e);
        }
```

**Rationale**: Using assertNull(null) in a catch block makes the test pass silently if the method being tested via reflection is renamed or removed, which can hide regressions.

File: src-test/src/com/etendoerp/metadata/auth/AuthUtilsGenerateTokenTest.java [hunk 9]
```diff
@@ -0,0 +1,326 @@
+package com.etendoerp.metadata.auth;
+
```
Comment (etendobot) [inline src-test/src/com/etendoerp/metadata/auth/AuthUtilsGenerateTokenTest.java:2]: > [!WARNING]
**⚠️ Blocking Issue**

Add the mandatory Etendo license header at the top of the file.

```java
/*
 *************************************************************************
 * The contents of this file are subject to the Etendo License
 * (the "License"), you may not use this file except in compliance
 * with the License.
 * You may obtain a copy of the License at
 * https://github.com/etendosoftware/etendo_core/blob/main/legal/Etendo_license.txt
 * Software distributed under the License is distributed on an
 * "AS IS" basis, WITHOUT WARRANTY OF ANY KIND, either express or
 * implied. See the License for the specific language governing rights
 * and limitations under the License.
 * All portions are Copyright (C) 2021-2025 FUTIT SERVICES, S.L
 * All Rights Reserved.
 * Contributor(s): Futit Services S.L.
 *************************************************************************
 */
package com.etendoerp.metadata.auth;
```

**Rationale**: New Java source files in com.etendoerp.* modules must include the standard Etendo license header. Since this is a new file, it is a blocking requirement.

File: src-test/src/com/etendoerp/metadata/auth/AuthUtilsGenerateTokenTest.java [hunk 10]
```diff
@@ -0,0 +1,326 @@
...
+            // Handle mock setup issues gracefully
+            assertNotNull(e);
+        }
+    }
+
+    /** Invokes Utils.decodeToken and asserts that the expected exception type is thrown. */
+    private void assertDecodeTokenThrowsExpectedException() {
+        try {
+            Utils.decodeToken("bad-token");
+        } catch (OBException e) {
```
Comment (etendobot) [inline src-test/src/com/etendoerp/metadata/auth/AuthUtilsGenerateTokenTest.java:270]: > [!WARNING]
**⚠️ Blocking Issue**

Add a fail() call immediately after the method invocation to ensure the test fails if no exception is thrown.

```java
            Utils.decodeToken("bad-token");
            org.junit.Assert.fail("Expected an exception to be thrown for bad-token");
```

**Rationale**: The current implementation of assertDecodeTokenThrowsExpectedException catches expected exceptions but does not fail if no exception is thrown. This means the test will pass even if the code under test is broken and fails to throw the expected exception.

File: src-test/src/com/etendoerp/metadata/auth/AuthUtilsGenerateTokenTest.java [hunk 11]
```diff
@@ -0,0 +1,326 @@
...
+    }
+
+    // ========== Helper Methods ==========
+
+    private void invokePrivateUtilsMethod(String methodName, Class<?>[] paramTypes, Object... args) {
+        try {
+            Method method = Utils.class.getDeclaredMethod(methodName, paramTypes);
+            method.setAccessible(true);
+            method.invoke(null, args);
+        } catch (Exception e) {
```
Comment (etendobot) [inline src-test/src/com/etendoerp/metadata/auth/AuthUtilsGenerateTokenTest.java:284]: **Suggestion**

Update the helper method to verify that an exception is actually thrown when expected.

```java
            method.invoke(null, args);
            // If an exception was expected, this point should not be reached.
            // Consider passing a boolean flag to the helper to indicate if failure is expected.
```

**Rationale**: The helper method invokePrivateUtilsMethod is used by tests that expect failure (e.g., testGetRoleThrowsOBExceptionOnNoSuchMethod). However, if method.invoke succeeds, the catch block is skipped and the test passes silently. This renders the 'failure' tests useless.

File: src-test/src/com/etendoerp/metadata/auth/AuthUtilsGenerateTokenTest.java [hunk 12]
```diff
@@ -0,0 +1,326 @@
...
+            swsUtilsMock = null;
+            swsUtilsMock = mockStatic(SecureWebServicesUtils.class);
+
+            // When decodeToken throws, it should wrap in OBException
+            swsUtilsMock.when(() -> SecureWebServicesUtils.decodeToken("bad-token"))
+                    .thenThrow(new RuntimeException("decode failed"));
+
+            assertDecodeTokenThrowsExpectedException();
+        } catch (Exception e) {
+            // Handle mock setup issues gracefully
```
Comment (etendobot) [inline src-test/src/com/etendoerp/metadata/auth/AuthUtilsGenerateTokenTest.java:261]: **Suggestion**

Do not catch exceptions in the test method unless you are specifically testing for them; let them propagate so the test framework marks the test as failed/errored.

```java
        } catch (Exception e) {
            org.junit.Assert.fail("Mock setup failed: " + e.getMessage());
        }
```

**Rationale**: Swallowing exceptions during mock setup and asserting only that the exception is not null causes the test to 'pass' even if the setup failed. Setup failures should cause the test to fail.

File: src-test/src/com/etendoerp/metadata/http/MetadataFilterAdditionalTest.java [hunk 13]
```diff
@@ -0,0 +1,507 @@
+package com.etendoerp.metadata.http;
+
```
Comment (etendobot) [inline src-test/src/com/etendoerp/metadata/http/MetadataFilterAdditionalTest.java:2]: > [!WARNING]
**⚠️ Blocking Issue**

Add the mandatory Etendo license header at the top of the file.

```java
/*
 *************************************************************************
 * The contents of this file are subject to the Etendo License
 * (the "License"), you may not use this file except in compliance
 * with the License.
 * You may obtain a copy of the License at
 * https://github.com/etendosoftware/etendo_core/blob/main/legal/Etendo_license.txt
 * Software distributed under the License is distributed on an
 * "AS IS" basis, WITHOUT WARRANTY OF ANY KIND, either express or
 * implied. See the License for the specific language governing rights
 * and limitations under the License.
 * All portions are Copyright (C) 2021-2024 FUTIT SERVICES, S.L
 * All Rights Reserved.
 * Contributor(s): Futit Services S.L.
 *************************************************************************
 */
package com.etendoerp.metadata.http;
```

**Rationale**: Every new Java class in a com.etendoerp.* module must include the standard Etendo license header as the first block comment in the file for legal and compliance reasons.

File: src-test/src/com/etendoerp/metadata/utils/CallAsyncProcessExtendedTest.java [hunk 14]
```diff
@@ -0,0 +1,453 @@
+package com.etendoerp.metadata.utils;
+
```
Comment (etendobot) [inline src-test/src/com/etendoerp/metadata/utils/CallAsyncProcessExtendedTest.java:2]: > [!WARNING]
**⚠️ Blocking Issue**

Add the mandatory Etendo license header at the top of the new file.

```java
/*
 *************************************************************************
 * The contents of this file are subject to the Etendo License
 * (the "License"), you may not use this file except in compliance
 * with the License.
 * You may obtain a copy of the License at
 * https://github.com/etendosoftware/etendo_core/blob/main/legal/Etendo_license.txt
 * Software distributed under the License is distributed on an
 * "AS IS" basis, WITHOUT WARRANTY OF ANY KIND, either express or
 * implied. See the License for the specific language governing rights
 * and limitations under the License.
 * All portions are Copyright (C) 2021-2024 FUTIT SERVICES, S.L
 * All Rights Reserved.
 * Contributor(s): Futit Services S.L.
 *************************************************************************
 */
package com.etendoerp.metadata.utils;
```

**Rationale**: Every new Java source file created within a com.etendoerp.* namespace must include the standard Etendo license header at the top of the file to comply with legal and formatting standards.

File: src-test/src/com/etendoerp/metadata/http/MetadataServletCoverageTest.java [hunk 15]
```diff
@@ -0,0 +1,318 @@
+package com.etendoerp.metadata.http;
+
```
Comment (etendobot) [inline src-test/src/com/etendoerp/metadata/http/MetadataServletCoverageTest.java:2]: > [!WARNING]
**⚠️ Blocking Issue**

Add the mandatory Etendo license header at the beginning of the file.

```java
/*
 *************************************************************************
 * The contents of this file are subject to the Etendo License
 * (the "License"), you may not use this file except in compliance
 * with the License.
 * You may obtain a copy of the License at
 * https://github.com/etendosoftware/etendo_core/blob/main/legal/Etendo_license.txt
 * Software distributed under the License is distributed on an
 * "AS IS" basis, WITHOUT WARRANTY OF ANY KIND, either express or
 * implied. See the License for the specific language governing rights
 * and limitations under the License.
 * All portions are Copyright (C) 2021-2026 FUTIT SERVICES, S.L
 * All Rights Reserved.
 * Contributor(s): Futit Services S.L.
 *************************************************************************
 */
package com.etendoerp.metadata.http;
```

**Rationale**: All new Java files in com.etendoerp.* modules must have the Etendo license header at the very top of the file.

File: src-test/src/com/etendoerp/metadata/http/MetadataServletCoverageTest.java [hunk 16]
```diff
@@ -0,0 +1,318 @@
...
+import java.io.PrintWriter;
+import java.io.StringWriter;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import com.etendoerp.metadata.exceptions.NotFoundException;
+import com.etendoerp.metadata.service.ServiceFactory;
+import com.etendoerp.metadata.utils.Utils;
```
Comment (etendobot) [inline src-test/src/com/etendoerp/metadata/http/MetadataServletCoverageTest.java:14]: **Suggestion**

Remove the unused import for ServiceFactory.

```java
// Remove unused import
```

**Rationale**: The ServiceFactory class is not explicitly used in the test code, only in comments. Unused imports should be removed to keep the code clean.

File: src-test/src/com/etendoerp/metadata/http/MetadataServletCoverageTest.java [hunk 17]
```diff
@@ -0,0 +1,318 @@
...
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.MockedStatic;
+import org.mockito.junit.MockitoJUnitRunner;
```
Comment (etendobot) [inline src-test/src/com/etendoerp/metadata/http/MetadataServletCoverageTest.java:20]: **Suggestion**

Remove the unused import for MockedStatic.

```java
// Remove unused import
```

**Rationale**: The MockedStatic class is imported but never used in the test methods. Unused imports should be removed.

File: src-test/src/com/etendoerp/metadata/http/NoteServletCoverageTest.java [hunk 18]
```diff
@@ -0,0 +1,420 @@
+package com.etendoerp.metadata.http;
+
```
Comment (etendobot) [inline src-test/src/com/etendoerp/metadata/http/NoteServletCoverageTest.java:2]: > [!WARNING]
**⚠️ Blocking Issue**

Add the mandatory Etendo license header to the new file.

```java
/*
 *************************************************************************
 * The contents of this file are subject to the Etendo License
 * (the "License"), you may not use this file except in compliance
 * with the License.
 * You may obtain a copy of the License at
 * https://github.com/etendosoftware/etendo_core/blob/main/legal/Etendo_license.txt
 * Software distributed under the License is distributed on an
 * "AS IS" basis, WITHOUT WARRANTY OF ANY KIND, either express or
 * implied. See the License for the specific language governing rights
 * and limitations under the License.
 * All portions are Copyright (C) 2021-2026 FUTIT SERVICES, S.L
 * All Rights Reserved.
 * Contributor(s): Futit Services S.L.
 *************************************************************************
 */
package com.etendoerp.metadata.http;
```

**Rationale**: New Java source files in the com.etendoerp.* namespace must include the standard Etendo license header at the very top.

File: src-test/src/com/etendoerp/metadata/http/NoteServletCoverageTest.java [hunk 19]
```diff
@@ -0,0 +1,420 @@
...
+/**
+ * Additional coverage tests for {@link NotesServlet}.
+ * Targets uncovered paths: exception handlers in doGet/doPost/doDelete,
+ * SWS path-based methods, isValidTable exception, empty request body,
+ * extractNoteIdFromPath edge cases, deleteNote exception, and
+ * canDeleteNote exception path.
+ */
+@RunWith(MockitoJUnitRunner.class)
+public class NoteServletCoverageTest extends OBBaseTest {
+
```
Comment (etendobot) [inline src-test/src/com/etendoerp/metadata/http/NoteServletCoverageTest.java:45]: **Suggestion**

Consider renaming the class to 'NotesServletCoverageTest' for consistency with 'NotesServlet'.

```java
public class NotesServletCoverageTest extends OBBaseTest {
```

**Rationale**: The servlet being tested is named 'NotesServlet' (plural). The test class name should be consistent with the class it tests.

File: src-test/src/com/etendoerp/metadata/utils/UtilsContextAndFieldTest.java [hunk 20]
```diff
@@ -0,0 +1,338 @@
+package com.etendoerp.metadata.utils;
+
```
Comment (etendobot) [inline src-test/src/com/etendoerp/metadata/utils/UtilsContextAndFieldTest.java:2]: > [!WARNING]
**⚠️ Blocking Issue**

Add the mandatory Etendo license header at the top of the file before the package declaration.

```java
/*
 *************************************************************************
 * The contents of this file are subject to the Etendo License
 * (the "License"), you may not use this file except in compliance
 * with the License.
 * You may obtain a copy of the License at
 * https://github.com/etendosoftware/etendo_core/blob/main/legal/Etendo_license.txt
 * Software distributed under the License is distributed on an
 * "AS IS" basis, WITHOUT WARRANTY OF ANY KIND, either express or
 * implied. See the License for the specific language governing rights
 * and limitations under the License.
 * All portions are Copyright (C) 2021-2024 FUTIT SERVICES, S.L
 * All Rights Reserved.
 * Contributor(s): Futit Services S.L.
 *************************************************************************
 */
package com.etendoerp.metadata.utils;
```

**Rationale**: Every new Java class in a com.etendoerp.* module must include the standard Etendo license header as the first block comment in the file to comply with legal and organizational standards.

File: src-test/src/com/etendoerp/metadata/utils/UtilsContextAndFieldTest.java [hunk 21]
```diff
@@ -0,0 +1,338 @@
...
+import static org.mockito.Mockito.mockStatic;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import javax.servlet.http.HttpServletRequest;
+
+import org.codehaus.jettison.json.JSONObject;
+import org.hibernate.criterion.Criterion;
+import org.openbravo.dal.service.OBCriteria;
+import org.junit.After;
```
Comment (etendobot) [inline src-test/src/com/etendoerp/metadata/utils/UtilsContextAndFieldTest.java:24]: **Suggestion**

Remove the duplicate import of OBCriteria on line 35.

```java

```

**Rationale**: The class 'org.openbravo.dal.service.OBCriteria' is already imported on line 22. Duplicate imports should be removed to keep the code clean.

File: src-test/src/com/etendoerp/metadata/service/MessageServiceTest.java [hunk 22]
```diff
@@ -234,4 +241,75 @@ public void testProcessMessagesWithMultipleParameters() throws IOException {
...
+    @Test
+    public void testSetCORSHeadersWithNullOrigin() {
+        HttpServletRequest req = mock(HttpServletRequest.class);
+        HttpServletResponse res = mock(HttpServletResponse.class);
+        when(req.getHeader(ORIGIN)).thenReturn(null);
+
+        messageService.setCORSHeaders(req, res);
+
+        verify(res, never()).setHeader(ACCESS_CONTROL_ALLOW_ORIGIN, null);
+    }
```
Comment (etendobot) [inline src-test/src/com/etendoerp/metadata/service/MessageServiceTest.java:258]: **Suggestion**

Use a generic argument matcher like any() to verify that the header is not set with any value when the origin is null.

```java
        verify(res, never()).setHeader(eq(ACCESS_CONTROL_ALLOW_ORIGIN), any());
```

**Rationale**: Verifying specifically for a null value is too restrictive. If the code incorrectly sets the header to any other value (like an empty string or a different origin), the test would still pass. Using any() ensures that the header is not set at all.

File: src-test/src/com/etendoerp/metadata/service/MessageServiceTest.java [hunk 23]
```diff
@@ -234,4 +241,75 @@ public void testProcessMessagesWithMultipleParameters() throws IOException {
...
+    @Test
+    public void testSetCORSHeadersWithEmptyOrigin() {
+        HttpServletRequest req = mock(HttpServletRequest.class);
+        HttpServletResponse res = mock(HttpServletResponse.class);
+        when(req.getHeader(ORIGIN)).thenReturn("");
+
+        messageService.setCORSHeaders(req, res);
+
+        verify(res, never()).setHeader(ACCESS_CONTROL_ALLOW_ORIGIN, "");
+    }
```
Comment (etendobot) [inline src-test/src/com/etendoerp/metadata/service/MessageServiceTest.java:273]: **Suggestion**

Use any() instead of an empty string in the never() verification to ensure the header is not set at all.

```java
        verify(res, never()).setHeader(eq(ACCESS_CONTROL_ALLOW_ORIGIN), any());
```

**Rationale**: Similar to the null origin case, verifying specifically for an empty string allows the test to pass even if the header is set to another value. It is better to verify that no call to setHeader with this key occurs.

File: src-test/src/com/etendoerp/metadata/service/MessageServiceTest.java [hunk 24]
```diff
@@ -234,4 +241,75 @@ public void testProcessMessagesWithMultipleParameters() throws IOException {
...
+    /**
+     * Tests process() when the error object is non-null.
+     * The response JSON should contain message, type, and title keys.
+     *
+     * @throws IOException if an I/O error occurs during processing
+     */
+    @Test
+    public void testProcessWithNonNullError() throws IOException {
+        when(mockRequest.getParameter("tabId")).thenReturn("test-tab-id");
+        responseWriter.getBuffer().setLength(0);
```
Comment (etendobot) [inline src-test/src/com/etendoerp/metadata/service/MessageServiceTest.java:303]: **Suggestion**

Mock the HttpSession and configure it to return a valid OBError object when getAttribute is called with the expected tab key.

```java
        when(mockRequest.getParameter("tabId")).thenReturn("test-tab-id");
        HttpSession session = mock(HttpSession.class);
        when(mockRequest.getSession(anyBoolean())).thenReturn(session);
        OBError error = new OBError();
        error.setMessage("Test Message");
        error.setType("Error");
        error.setTitle("Test Title");
        when(session.getAttribute("TABS|test-tab-id")).thenReturn(error);
```

**Rationale**: The test 'testProcessWithNonNullError' aims to verify the behavior when an error is present. However, it does not set up an OBError in the session. Since VariablesSecureApp retrieves messages from the session, vars.getMessage(tabId) will return null in the current setup, and the test will actually execute the null-error branch.

File: src/com/etendoerp/metadata/builders/FieldBuilder.java [hunk 25]
```diff
@@ -661,6 +678,46 @@ public static String getPropertyOrDataSourceField(SelectorField selectorField) {
...
+     * @param selectorField The selector field to resolve the reference ID for
+     * @return The reference ID string, or null if none is found
+     */
+    private static String resolveReferenceId(SelectorField selectorField) {
+        if (selectorField.getReference() != null) {
+            return selectorField.getReference().getId();
+        }
+        if (selectorField.getColumn() != null) {
+            return selectorField.getColumn().getReference().getId();
+        }
```
Comment (etendobot) [inline src/com/etendoerp/metadata/builders/FieldBuilder.java:694]: > [!WARNING]
**⚠️ Blocking Issue**

Add a null check for the column's reference before accessing its ID.

```java
            return selectorField.getColumn().getReference() != null ? selectorField.getColumn().getReference().getId() : null;
```

**Rationale**: Accessing .getId() directly on the result of getReference() without a null check can lead to a NullPointerException if the data is inconsistent or not fully loaded, violating the rule against unsafe property access.

File: src/com/etendoerp/metadata/builders/FieldBuilder.java [hunk 26]
```diff
@@ -310,6 +310,23 @@ protected static JSONObject addSelectorInfo(String fieldId, Selector selector) t
         selectorInfo.put(Constants.DISPLAY_FIELD_PROPERTY, getDisplayField(selector));
         selectorInfo.put(Constants.VALUE_FIELD_PROPERTY, getValueField(selector));
 
+        if (selector.getProcessDefintion() != null) {
+            selectorInfo.put("hasProcessDefinitionRelated", true);
+            selectorInfo.put("processDefinitionId", selector.getProcessDefintion().getId());
```
Comment (etendobot) [inline src/com/etendoerp/metadata/builders/FieldBuilder.java:315]: **Suggestion**

Consider using a constant for the JSON key "hasProcessDefinitionRelated" to maintain consistency with other parts of the Metadata Builders.

```java
            selectorInfo.put(Constants.HAS_PROCESS_DEFINITION_RELATED, true);
```

**Rationale**: Using string literals for JSON keys is prone to typos and makes maintenance harder. It is recommended to use constants for property keys, as seen elsewhere in the class (e.g., Constants.DISPLAY_FIELD_PROPERTY).

File: src/com/etendoerp/metadata/utils/LegacyUtils.java [hunk 27]
```diff
@@ -103,23 +104,24 @@ public static boolean isMutableSessionAttribute(String attribute) {
     }
 
     /**
-     * Finds the ID of the first active tab that belongs to the given window and table.
+     * Finds the ID of the first active tab that belongs to the given window and
+     * table.
      * The result is ordered by sequence number to ensure deterministic output when
```
Comment (etendobot) [inline src/com/etendoerp/metadata/utils/LegacyUtils.java:109]: **Suggestion**

Keep the Javadoc description on a single line.

```java
     * Finds the ID of the first active tab that belongs to the given window and table.
```

**Rationale**: The Javadoc description was unnecessarily split into two lines. Keeping it on a single line improves readability and adheres to standard formatting when the line length is within limits.

File: src/com/etendoerp/metadata/utils/LegacyUtils.java [hunk 28]
```diff
@@ -103,23 +104,24 @@ public static boolean isMutableSessionAttribute(String attribute) {
...
     public static String findTabIdByWindowAndTable(String windowId, String tableId) {
         if (StringUtils.isEmpty(windowId) || StringUtils.isEmpty(tableId)) {
             return null;
         }
         OBQuery<Tab> query = OBDal.getInstance().createQuery(
-            Tab.class,
-            "where window.id = :windowId and table.id = :tableId and active = true order by sequenceNumber"
-        );
+                Tab.class,
+                "where window.id = :windowId and table.id = :tableId and active = true order by sequenceNumber");
```
Comment (etendobot) [inline src/com/etendoerp/metadata/utils/LegacyUtils.java:124]: **Suggestion**

Use a 4-space continuation indent for method parameters.

```java
            Tab.class,
```

**Rationale**: The indentation for method arguments was increased from 4 spaces to 8 spaces relative to the base indentation. Etendo standards typically use a 4-space continuation indent.

**Review by etendobot — CHANGES_REQUESTED** [review]
Changes requested by agent. Please resolve blocking issues.
